core/base/arrayPreconditioning/ArrayPreconditioning.h

```diff
@@ -0,0 +1,593 @@
+                        const DT *const scalars,
+                        const IT *const globalIds,
+                        const char *const ghostCells) const {
+      for(size_t i = 0; i < nVerts; i++) {
+        IT globalId = globalIds[i];
+        IT localId = i;
+        gidToLidMap[globalId] = localId;
+        if((int)ghostCells[i] == 0) {
+          float scalarValue = scalars[i];
+          valuesToSortVector.emplace_back(scalarValue, globalId, localId);
+        } else {
+          gidsToGetVector.push_back(globalId);
+        }
+      }
+    }
+
+    // orders an value vector first by their scalar value and then by global id
+    template <typename DT, typename IT>
+    void sortVerticesDistributed(std::vector<value<DT, IT>> &values) const {
+      std::sort(
+        values.begin(), values.end(), [](value<DT, IT> v1, value<DT, IT> v2) {
+          return (v1.scalar < v2.scalar)
+                 || (v1.scalar == v2.scalar && v1.globalId < v2.globalId);
+        });
+    }
+
+    // send the highest burstSize values and decrease the vector by that amount
+    // check if there are actually that many elements in the vector
+    template <typename DT, typename IT>
+    void returnVectorForBurstsize(std::vector<value<DT, IT>> &outVector,
+                                  std::vector<value<DT, IT>> &values,
+                                  size_t burstSize) const {
+
+      if(burstSize > values.size()) {
+        outVector.assign(values.begin(), values.end());
+        values.clear();
+      } else {
+        outVector.assign(values.end() - burstSize, values.end());
+        values.erase(values.end() - burstSize, values.end());
+      }
+    }
+
+    /**
+     * @brief Sort vertices according to scalars disambiguated by offsets
+     *
+     * @param[in] nInts number of long ints, as globalid / order pairs
+     * @param[in] orderedValuesForRank array of size nInts, the actual pairs,
+     * [i] = gId, [i+1] = order
+     * @param[in] gidToLidMap map which maps scalar values to a defined order
+     * @param[out] order array of size nVerts, computed order of vertices, this
+     * procedure doesn't fill it completely
+     * @param[in] nThreads number of parallel threads
+     */
+    template <typename IT>
+    void buildArrayForReceivedData(const size_t nInts,
+                                   const IT *const orderedValuesForRank,
+                                   std::unordered_map<IT, IT> &gidToLidMap,
+                                   SimplexId *const order,
+                                   const int nThreads) const {
+
+      TTK_FORCE_USE(nThreads);
+
+#ifdef TTK_ENABLE_OPENMP
+#pragma omp parallel for num_threads(nThreads)
+#endif // TTK_ENABLE_OPENMP
+      for(size_t i = 0; i < nInts; i += 2) {
+        order[gidToLidMap[orderedValuesForRank[i]]]
+          = orderedValuesForRank[i + 1];
+      }
+    }
+
+    // gets all the neighbors of a rank based on the rankarray of the vertices
```

Comment:
NOTE for later: that function should be a member of all the classes deriving from the BaseClass (if MPI is detected and enabled).